### PR TITLE
01 feat: add shared VAE infrastructure 

### PR DESCRIPTION
**Pull request by ryanontheinside**
Add `base/vae` module with factory pattern for VAE selection.

- New VAE implementations: `KreaRealtimeVideoVAE`, `LongLiveVAE`, `StreamDiffusionV2VAE`
- Factory function `create_vae()` for strategy-based instantiation
- Shared constants for latent normalization

Purely additive - no existing code modified.

Will be utilized in following PRs.

### CODE REVIEW

File: src/scope/core/pipelines/base/vae/streamdiffusionv2.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+        output = self.model.stream_decode(zs, scale).float().clamp_(-1, 1)
+        output = output.permute(0, 2, 1, 3, 4)
+        return output
+
+    def clear_cache(self):
+        """Clear decoder cache for next sequence."""
+        self.model.first_batch = True
+
+
+class StreamDiffusionV2VAEWithLongLiveScaling(StreamDiffusionV2VAE):
```
Comment (ryanontheinside) [inline src/scope/core/pipelines/base/vae/streamdiffusionv2.py]: I'm not overly thrilled with this. As I work on LightVAE I find that I will have to do the same thing, subclass it for longlive's scaling. Then I imagine a world where we add additional models with further scaling requirements. A couple of alternative solutions might be:

- Callback approach: Pass a callable encoding_scaling_fn (or None). Define standard functions like longlive_encoding_scaling() in the module. 

- Enum + registry: EncodingScalingMode enum with a registry mapping to functions. More discoverable and type-safe, but adds complexity

Comment (leszko) [reply src/scope/core/pipelines/base/vae/streamdiffusionv2.py]: Does it mean that `StreamDiffusionV2VAE` does not work with LongLive? If yes, then it makes me think that VAE are not interchangeable between different pipelines. If that's the case, I wonder if we couldn't "make" them interchangeable, so that each VAE is possible to apply to every pipeline. Some ideas:
1. (if possible) Detect (here or in the related block) if the normalization is required and if yes, then apply it; Then we would have only `StreamDiffusionV2VAE`
2. Have a param (here or in the related block) to decide if the normalization should happen

File: src/scope/core/pipelines/base/vae/krea_realtime_video.py [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+        self.encoder = VAEEncoderWrapper(vae)
+        self.encoder.eval()
+        self.encoder.requires_grad_(False)
+
+        self.decoder = VAEDecoderWrapper()
+        state_dict = torch.load(vae_path, map_location="cpu")
+        decoder_state_dict = {}
+        for key, value in state_dict.items():
+            if "decoder." in key or "conv2" in key:
+                decoder_state_dict[key] = value
```
Comment (leszko) [inline src/scope/core/pipelines/base/vae/krea_realtime_video.py]: What is this code block doing? The variable `decoder_state_dict` does not seem to be used anywhere.

File: src/scope/core/pipelines/base/vae/krea_realtime_video.py [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+    This VAE uses a specialized architecture optimized for real-time video
+    generation with cached encoding and decoding.
+    """
+
+    def __init__(
+        self,
+        model_name: str = "Wan2.1-T2V-1.3B",
+        model_dir: str | None = None,
+        vae_path: str | None = None,
+    ):
```
Comment (leszko) [inline src/scope/core/pipelines/base/vae/krea_realtime_video.py]: I wonder if we should maybe make all the VAE constructors looking the same. I see that in this one you can define `model_name`. `model_dir`, and `vae_path`. At the same time in `LongLiveVAE` you have only 1 parameter `model_dir` and the rest is hardcoded.

File: src/scope/core/pipelines/base/vae/krea_realtime_video.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+        """
+        output, self.decoder_cache = self.decoder(latent, *self.decoder_cache)
+        return output
+
+    def clear_cache(self):
+        """Clear decoder cache for next sequence."""
+        self.decoder_cache = [None] * 55
+
+
+class VAEEncoderWrapper(nn.Module):
```
Comment (leszko) [inline src/scope/core/pipelines/base/vae/krea_realtime_video.py]: Sorry, super stupid question, why for the Krea we need to write our own wrapper while for other pipelines we use the already implemented `WanVAE_()`?

And super nit: if this wrapper is potentially reusable, I'd move it out of this file.

**Review by leszko — COMMENTED** [review]
Very nice PR @ryanontheinside . Added some comments. My main points are:

1. I asked this elsewhere already, so apologies if you’ve responded before.) Just to confirm: we want to move all VAEs into the `pipelines/base/vae` directory instead of keeping them alongside their respective pipelines, right? I’m fine with that, but it does feel a bit counter-intuitive - especially as we continue adding more pipelines. That’s why I wanted to double-check.
2. My understanding is that the VAEs aren’t currently interchangeable across pipelines - for example, LongLive requires latent normalization, and Krea uses its own VAE wrapper. Is that correct? And if so, do we eventually want to make the VAEs interchangeable?

File: src/scope/core/pipelines/base/vae/streamdiffusionv2.py (hunk 1 above)
Comment (ryanontheinside) [reply src/scope/core/pipelines/base/vae/streamdiffusionv2.py]: > Does it mean that `StreamDiffusionV2VAE` does not work with LongLive? If yes, then it makes me think that VAE are not interchangeable between different pipelines.

Correct - StreamDiffusionV2VAE doesn't work directly with LongLive because it doesn't apply the latent normalization that LongLive expects. The goal is to make VAEs interchangeable when possible - this strategy pattern is a first step, and the subclass for LongLive scaling is a stopgap until we decide on a better approach. I prefer the callback approach I mentioned earlier, since other models may have different scaling requirements. I expect we'll see other common patterns across VAE implementations as we add new models and VAEs.

File: src/scope/core/pipelines/base/vae/krea_realtime_video.py (hunk 4 above)
Comment (ryanontheinside) [reply src/scope/core/pipelines/base/vae/krea_realtime_video.py]: > Sorry, super stupid question, why for the Krea we need to write our own wrapper while for other pipelines we use the already implemented `WanVAE_()`?
> 
> And super nit: if this wrapper is potentially reusable, I'd move it out of this file.

We need the wrappers because unlike the other models Krea splits the encoder and decoder with separate caches. They aren't reusable elsewhere yet, so we could leave them for now.

File: src/scope/core/pipelines/base/vae/krea_realtime_video.py (hunk 3 above)
Comment (ryanontheinside) [reply src/scope/core/pipelines/base/vae/krea_realtime_video.py]: > I wonder if we should maybe make all the VAE constructors looking the same.

I agree we should standardize the constructors

File: src/scope/core/pipelines/base/vae/krea_realtime_video.py (hunk 2 above)
Comment (ryanontheinside) [reply src/scope/core/pipelines/base/vae/krea_realtime_video.py]: > What is this code block doing? The variable `decoder_state_dict` does not seem to be used anywhere.

this is a bug - the variable is vestigial and never used. I'll fix it.

File: src/scope/core/pipelines/wan2_1/vae/constants.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
+"""Shared VAE normalization constants for Wan2.1 models.
```
Comment (yondonfu) [inline src/scope/core/pipelines/wan2_1/vae/constants.py:1]: The fact that this is specific to Wan2.1 makes me feel like this should be under `wan2_1`.

File: src/scope/core/pipelines/base/vae/krea_realtime_video.py [hunk 6]
```diff
@@ -0,0 +1,209 @@
+"""KreaRealtimeVideo VAE wrapper with encoder/decoder separation."""
+
+import os
+
+import torch
+import torch.nn as nn
+
+from ...krea_realtime_video.modules.vae import CausalConv3d, WanVAE
```
Comment (yondonfu) [inline src/scope/core/pipelines/base/vae/krea_realtime_video.py]: As a convention, IMO we should avoid ... imports and only do .. imports. If a ... is needed an absolute import could be clearer eg `from scope.core.pipelines.krea_realtime_video...`. WDYT about this convention? We should just decide on one and stick with it.

File: src/scope/core/pipelines/wan2_1/vae/constants.py (hunk 5 above)
Comment (ryanontheinside) [reply src/scope/core/pipelines/wan2_1/vae/constants.py:1]: i agree

File: src/scope/core/pipelines/base/vae/krea_realtime_video.py (hunk 6 above)
Comment (ryanontheinside) [reply src/scope/core/pipelines/base/vae/krea_realtime_video.py]: works for me

**Comment by ryanontheinside** [comment]
> I'm wondering whether we can avoid having to maintain all of these VAE comp/module impls.
> 
> AFAICT (not the most thorough review yet) the impls are very similar minus a few highlights:
> 
> * krea-realtime-video uses an external cache that is always expected to be passed in and the impl returns the updated cache along with results
> * streamdiffusionv2 is missing a normalization step during encoding that the other impls have - should that just be there always?
> 
> I created this [spike](https://github.com/daydreamlive/scope/commit/ae772be997ac7cda07966dac252179f0ddeb74e5#r171740317) where every pipeline uses a modified version of the streamdiffusionv2 VAE comp/module impls and while I haven't tested thoroughly/benchmarked properly it does like all the pipelines work without any major noticeable regression. And streamdiffusionv2 output looks...different (possibly better?) with the added normalization step during encoding.

Your spike validates consolidating the Wan VAE implementations - agree we should do that, much more simple and elegant! I thought the normalization being missing from sd was intentional but this simplifies things further. Excited to try it as I would assume it will improve the quality of the output too.

The factory pattern will still be useful for LightVAE which is a fundamentally different model but same interface. See the WIP branch: #208 
Im thinking plan:
This PR: 
- Consolidate Wan VAEs into one implementation (per your spike)
- Move to wan directory
- Keep factory
Follow-up PR: Add LightVAE as a factory option

WDYT?

File: src/scope/core/pipelines/wan2_1/vae/wan.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+import os
+
+import torch
+
+from scope.core.pipelines.streamdiffusionv2.modules.vae import _video_vae
+
+from .constants import WAN_VAE_LATENT_MEAN, WAN_VAE_LATENT_STD
+
+
+class WanVAE(torch.nn.Module):
```
Comment (yondonfu) [inline src/scope/core/pipelines/wan2_1/vae/wan.py]: Let's keep the *Wrapper naming convention for now?

File: src/scope/core/pipelines/wan2_1/vae/wan.py [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+        device, dtype = pixel.device, pixel.dtype
+        scale = self._get_scale(device, dtype)
+
+        if use_cache:
+            # Streaming encode - cache is maintained across calls
+            latent = self.model.stream_encode(pixel)
+            # Apply normalization (stream_encode returns unnormalized)
+            latent = self._apply_encoding_normalization(latent, scale)
+        else:
+            # Batch encode - cache is cleared, normalization applied internally
```
Comment (yondonfu) [inline src/scope/core/pipelines/wan2_1/vae/wan.py]: We don't want to clear the cache for Krea because the first frame re-encode shouldn't affect the cache used for streaming encodes. The way I handled this in the spike branch I shared was to manually pass in a one-time use cache which should work, but open to other solutions.

File: src/scope/core/pipelines/wan2_1/vae/wan.py [hunk 9]
```diff
@@ -0,0 +1,127 @@
+"""Unified Wan VAE wrapper with streaming and batch encoding/decoding."""
+
+import os
+
+import torch
+
+from scope.core.pipelines.streamdiffusionv2.modules.vae import _video_vae
```
Comment (yondonfu) [inline src/scope/core/pipelines/wan2_1/vae/wan.py]: Can we copy the module under `wan2_1/vae/modules`? Then, we can remove the per pipeline VAE modules in a later PR.

File: src/scope/core/pipelines/wan2_1/vae/wan.py (hunk 7 above)
Comment (yondonfu) [reply src/scope/core/pipelines/wan2_1/vae/wan.py]: Bump.

File: src/scope/core/pipelines/wan2_1/vae/wan.py (hunk 8 above)
Comment (yondonfu) [reply src/scope/core/pipelines/wan2_1/vae/wan.py]: Bump - do you intend to handle this separately? If so, just confirm.

**Review by yondonfu — CHANGES_REQUESTED** [review]

File: src/scope/core/pipelines/wan2_1/vae/wan.py (hunk 8 above)
Comment (ryanontheinside) [reply src/scope/core/pipelines/wan2_1/vae/wan.py]: i had,  but I am rebuilding branch 3 so better to do it here

File: src/scope/core/pipelines/wan2_1/vae/wan.py (hunk 7 above)
Comment (ryanontheinside) [reply src/scope/core/pipelines/wan2_1/vae/wan.py]: My bad - i totally missed this one
